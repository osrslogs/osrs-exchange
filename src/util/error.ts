/* eslint-disable max-classes-per-file */

/**
 * A Error with data details. `data` will containing the data
 * that made the error occur.
 */
class DataError extends Error {
  data: string;

  constructor(msg: string, data: string) {
    super(msg);
    this.data = data;
  }
}

/**
 * Error used when a html page has unexpected structure, e.g: Could not parse element.
 * The raw html can be found in the `data` property.
 */
export class InvalidHtmlError extends DataError {}

/**
 * A generic HTTP error, default status 500.
 * Used to map HTTP status codes to errors.
 */
export class HttpError extends Error {
  status: number;

  constructor(msg: string, status = 500) {
    super(msg);
    this.status = status;
  }
}

export class NotFoundError extends HttpError {
  constructor(msg: string) {
    super(msg, 403);
  }
}

export class ServiceUnavailableError extends HttpError {
  constructor(msg: string) {
    super(msg, 503);
  }
}
