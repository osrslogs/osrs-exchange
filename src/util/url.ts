const BASE_URL = 'http://services.runescape.com/m=itemdb_oldschool';

export const buildItemsPageUrl = (alpha: string, page: number): string => {
  return `${BASE_URL}/api/catalogue/items.json?category=1&alpha=${alpha}&page=${page}`;
};

export const buildItemPriceGraphUrl = (id: number): string => {
  return `${BASE_URL}/api/graph/${id}.json`;
};

/**
 * Builds the item url for the given item id. Note: It will redirect 302->200
 * due to not providing the name in the URL.
 *
 * @param {number} id The id of the item to lookup
 *
 * @returns {string} The url
 */
export const buildItemUrl = (id: number): string => {
  return `${BASE_URL}/viewitem?obj=${id}`;
};
