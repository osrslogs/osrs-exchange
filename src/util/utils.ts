import jsonItems from './data/items.json';
import { JsonItem, ItemName } from '../types';

const items: JsonItem = jsonItems as JsonItem;

/*
const getKeyByValue = (obj: object, value: unknown): unknown => {
  return Object.keys(obj).find(key => obj[key] === value);
};
*/

/**
 * Lookups the item id of an item by name
 *
 * @param {ItemName} name The item name
 *
 * @returns {number | undefined} The item id or undefined if not found
 */
export const getItemId = (name: ItemName): number | undefined => {
  return items[name.toLowerCase()];
};

/**
 * Converts a string, with assumed numbers, to number by removing
 * spaces and commas.
 *
 * @param {string} number
 *
 * @returns {number}
 */
export const cleanNumber = (number: string): number => {
  return Number(number.replace(/[ ,]+/g, ''));
};
