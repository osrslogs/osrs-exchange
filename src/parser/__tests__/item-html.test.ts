import parseItemHtml from '../item-html';
import htmlConfig from '../../util/test-util/html-mock.json';
import { InvalidHtmlError } from '../../util/error';
import { Item } from '../../types';

describe('parseItemHtml', () => {
  it('throws error on invalid page', () => {
    expect(() => parseItemHtml(htmlConfig.invalidPage)).toThrow(InvalidHtmlError);
  });

  it('it returns Item when given a valid page', () => {
    const item: Item = parseItemHtml(htmlConfig.validPage);
    expect(item).toMatchSnapshot();
  });
});
