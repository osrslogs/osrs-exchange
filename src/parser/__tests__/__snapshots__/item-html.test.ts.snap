// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseItemHtml it returns Item when given a valid page 1`] = `
Object {
  "description": "A weapon from the abyss.",
  "icon": "http://services.runescape.com/m=itemdb_oldschool/1583157682291_obj_big.gif?id=4151",
  "id": 4151,
  "members": true,
  "name": "Abyssal whip",
  "price": Object {
    "change": Object {
      "day180": -5981,
      "day30": -5981,
      "day90": -5981,
      "today": -5981,
    },
    "current": 2388108,
  },
}
`;
